File: src/ios/OSInAppBrowserError.swift
enum OSInAppBrowserTarget {
    case externalBrowser
    case systemBrowser
    case webView
}

enum OSInAppBrowserError: Error {
    case inputArgumentsIssue(target: OSInAppBrowserTarget)
    case failedToOpen(url: String, onTarget: OSInAppBrowserTarget)
    case noBrowserToClose
    
    private var code: Int {
        return switch self {
        case .inputArgumentsIssue(let target):
            switch target {
            case .externalBrowser: 5
            case .systemBrowser: 6
            case .webView: 7
            }
        case .failedToOpen(url: _, onTarget: let target):
            switch target {
            case .externalBrowser: 8
            case .systemBrowser: 9
            case .webView: 11
            }
        case .noBrowserToClose: 12
        }
    }
    
    private var description: String {
        let result: String
        
        switch self {
        case .inputArgumentsIssue(let target):             
            let targetString = switch target {
            case .externalBrowser: "openInExternalBrowser"
            case .systemBrowser: "openInSystemBrowser"
            case .webView: "openInWebView"
            }
            
            result = "The '\(targetString)' input parameters aren't valid."
        case .failedToOpen(url: let url, onTarget: let target):
            let targetString = switch target {
            case .externalBrowser: "External browser"
            case .systemBrowser: "SafariViewController"
            case .webView: "The WebView"
            }
            
            result = "\(targetString) couldn't open the following URL: '\(url)'"
        case .noBrowserToClose:
            result = "There’s no browser view to close."
        }
        
        return result
    }
    
    func toDictionary() -> [String: String] {
        [
            "code": "OS-PLUG-IABP-\(String(format: "%04d", self.code))",
            "message": self.description
        ]
    }   
}

File: src/android/OSInAppBrowserError.kt
package com.outsystems.plugins.inappbrowser.osinappbrowser

enum class OSInAppBrowserError(val code: Int, val message: String) {
    INPUT_ARGUMENTS_EXTERNAL_BROWSER_ISSUE(100, "The 'openInExternalBrowser' input parameters aren't valid."),
    INPUT_ARGUMENTS_SYSTEM_BROWSER_ISSUE(101, "The 'openInSystemBrowser' input parameters aren't valid."),
    INPUT_ARGUMENTS_WEB_VIEW_ISSUE(102, "The 'openInWebView' input parameters aren't valid."),
    OPEN_EXTERNAL_BROWSER_FAILED(103, "External browser couldn't open the following URL: {url}"),
    OPEN_SYSTEM_BROWSER_FAILED(104, "Custom Tabs couldn't open the following URL: {url}"),
    OPEN_WEB_VIEW_FAILED(105, "The WebView couldn't open the following URL: {url}"),
    CLOSE_FAILED(106, "There’s no browser view to close.");

    fun getErrorMessage(url: String?): String {
        return url?.let {
            message.replace("{url}", url)
        } ?: message
    }

    fun formatErrorCode(): String {
        return "OS-PLUG-IABP-" + code.toString().padStart(4, '0')
    }

}